src/storage/LockingResourceStore.ts

```diff
@@ -55,4 +54,24 @@ export class LockingResourceStore implements AtomicResourceStore {
       await lock.release();
     }
   }
+
+  private async lockedRunGet(identifier: ResourceIdentifier, preferences: RepresentationPreferences,
+    conditions?: Conditions): Promise<Representation> {
+    const lock = await this.locks.acquire(identifier);
+
+    // Execute the getRepresentation function and return the resulting Representation.
+    let result;
+    try {
+      result = await this.source.getRepresentation(identifier, preferences, conditions);
+      return result;
+    } finally {
+      // Wait for the end or error event to be called to release the lock if the result contains a valid Readable.
+      if (!result || !result.data) {
+        await lock.release();
+      } else {
+        result.data.on('end', async(): Promise<void> => lock.release());
```

Comment:
Looking at https://nodejs.org/api/stream.html#stream_class_stream_readable, I think we need to listen to `close`.
From that page:

- The 'end' event will not be emitted unless the data is completely consumed. 
- `readable.destroy([error])` Optionally emit an 'error' event, and emit a 'close' event (unless emitClose is set to false). 

But also:

- A Readable stream will always emit the 'close' event if it is created with the emitClose option.

While it seems that `emitClose` is `true` by default, might still be wise to listen to both `end` and `close` (also because `end` comes sooner, if emitted, so faster lock release).

Reply:
In any case, we need a test for a destroyed stream.

Reply:
Side remark: I also looked at the `close` event, but since for example AsyncIterator doesn't emit one as far as I know I wondered how safe it was to depend on streams emitting that event?

Reply:
It seems that one of `error, `end`, or `close` will always be emitted (unless a stream with `emitClose` set to `false` is destroyed).

AsyncIterator should perhaps do this; could you investigate https://github.com/RubenVerborgh/AsyncIterator/issues/21 ?